utils/run-clang-tidy.py

```diff
@@ -122,33 +131,302 @@ def make_absolute(f, directory):
     return os.path.normpath(os.path.join(directory, f))
 
 
-def get_tidy_invocation(f, clang_tidy_binary, checks, tmpdir, build_path,
-                        header_filter, extra_arg, extra_arg_before, quiet,
-                        config):
-    """Gets a command line for clang-tidy."""
-    start = [clang_tidy_binary]
+cache = None
+
+class Cache:
+    """Implements the cache.
+
+    Acknowledgements:
+
+    The basic inspiration and some ideas for hashing the sources came from
+    https://github.com/freedick/cltcache.
+
+    Some more ideas for hashing and ideas for parsing compile_commands.json
+    came from https://github.com/matus-chochlik/ctcache.
+
+    No code is copied directly from either.
+    """
+    failed_files_regexp = re.compile(r'^([0-9a-f]{64})\.std(out|err)$')
+
+    def __init__(self, dir, tidy_version, config, checks, custom):
+        self.dir = os.path.join(dir, 'clang-tidy-cache')
+        if not os.path.isdir(self.dir):
+            os.mkdir(self.dir)
+
+        self.config_hash = get_hash(tidy_version + config + checks)
+
+        # We only put caches of custom runs in separate subdirs, so we can clean
+        # up the main directory on upgrades or when the default config changes.
+        # Custom subdirs have to be cleaned up manually on upgrades.
+        if custom:
+            self.dir = os.path.join(self.dir, '__custom_args', self.config_hash)
+            if not os.path.isdir(self.dir):
+                os.mkdir(self.dir)
+
+        self.passed_file = os.path.join(self.dir, 'passed')
+        self.failed_dir = os.path.join(self.dir, 'failed')
+        if not os.path.isdir(self.dir):
+            os.mkdir(self.dir)
```

Comment:
This is already done above.
```suggestion
```

Reply:
Oops, must have been left there when I added the custom dir handling code.